bus-mapping/src/circuit_input_builder.rs

```diff
@@ -255,54 +310,75 @@ impl<'a> CircuitInputBuilder {
-    /// [`OperationRef`](crate::exec_trace::OperationRef) to each of the
-    /// generated operations.
-    fn handle_tx(
-        &mut self,
-        eth_tx: &eth_types::Transaction,
-        geth_trace: &GethExecTrace,
-        is_last_tx: bool,
-    ) -> Result<(), Error> {
-        let mut tx = self.new_tx(eth_tx, !geth_trace.failed)?;
-        let mut tx_ctx = TransactionContext::new(eth_tx, geth_trace, is_last_tx)?;
-
-        // Generate BeginTx step
-        let begin_tx_step = gen_associated_steps(
-            &mut self.state_ref(&mut tx, &mut tx_ctx),
-            ExecState::BeginTx,
-        )?;
-        tx.steps_mut().push(begin_tx_step);
-
-        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
-            let mut state_ref = self.state_ref(&mut tx, &mut tx_ctx);
-            log::trace!("handle {}th opcode {:?} ", index, geth_step.op);
-            let exec_steps = gen_associated_ops(
-                &geth_step.op,
-                &mut state_ref,
-                &geth_trace.struct_logs[index..],
-            )?;
-            tx.steps_mut().extend(exec_steps);
+impl CircuitInputBuilder<DynamicCP> {
+    /// Handle a block by handling each transaction to generate all the
+    /// associated operations. From these operations, the optimal circuit parameters
+    /// are derived and set.
+    pub fn handle_block(
+        mut self,
+        eth_block: &EthBlock,
+        geth_traces: &[eth_types::GethExecTrace],
+    ) -> Result<CircuitInputBuilder<ConcreteCP>, Error> {
+        // accumulates gas across all txs in the block
+        for (tx_index, tx) in eth_block.transactions.iter().enumerate() {
+            let geth_trace = &geth_traces[tx_index];
+            self.handle_tx(tx, geth_trace, tx_index + 1 == eth_block.transactions.len())?;
         }
+        self.set_value_ops_call_context_rwc_eor();
+        let c_params = {
+            let max_txs = eth_block.transactions.len();
+            let max_bytecode = self.code_db.0.values().fold(0, |acc, a| acc + a.len() + 1);
+
+            let max_calldata = eth_block
+                .transactions
+                .iter()
+                .fold(0, |acc, tx| acc + tx.input.len());
+            let max_exp_steps = self.block.exp_events.len();
+            // The `+ 2` is used to take into account the two extra empty copy rows needed
+            // to satisfy the query at `Rotation(2)` performed inside of the
+            // `rows[2].value == rows[0].value * r + rows[1].value` requirement in the RLC
+            // Accumulation gate.
+            let max_copy_rows = self
+                .block
+                .copy_events
+                .iter()
+                .fold(0, |acc, c| acc + c.bytes.len())
+                * 2
+                + 2;
+            let max_rws: usize = self.block_ctx.rwc.into();
+            let max_evm_rows = 0;
+            let max_keccak_rows = 0;
+            ConcreteCP {
+                // TODO Re-check RW padding warning
```

Comment:
I think the RwTable requires some extra rows (I think it may be just 1).  Padding rows in the RwTable have the Tag=Start, and we do 2 lookups to Start rows from EVM Circuit to guarantee the total number of rw entries.  I think both lookups could point to the same row, that's why I think it's just 1, but maybe I forgot something.

On the other hand... with this https://github.com/privacy-scaling-explorations/halo2/pull/185 we could find a better approach. But that's a topic for another task!

Reply:
I forgot to reply this!  So in the current merged state with `max_rws: max_rws + 1` there's a warning right?  If that's the case, could you open an issue to track this?  You can copy a link to this conversation in the issue :)